File: src/todo.hs
import System.Environment
import qualified Data.ByteString.Lazy.Char8 as B
import System.IO
import System.IO.Error
import System.Directory
import Data.List

dispatch :: [(String, [String] -> IO())]
dispatch = [("add", add),
            ("view", view),
            ("del", del)
           ]

main = catchIOError doWork eHandler

doWork = do
    (cmd:args) <- getArgs
    let (Just action) = lookup cmd dispatch
    action args

eHandler :: IOError -> IO()
eHandler e 
    | isDoesNotExistError e = 
        case ioeGetFileName e of Just path -> putStrLn $ path ++ " doesn't exist."
                                 Nothing -> putStrLn "File doesn't exist."
    | otherwise = ioError e

add :: [String] -> IO()
add [] = putStrLn "Nothing to add."
add [item] = appendFile "todo.txt" (item ++ "\n")

view :: [String] -> IO()
view [] = do
    content <- readFile "todo.txt"
    putStr $ (unlines . zipWith (\n line -> show n ++ " - " ++ line) [0..] . lines) content

del :: [String] -> IO()
del [] = putStrLn "Nothing to delete."
del [itemId] = do
    content <- B.readFile "todo.txt"
    (tempName, tempHandle) <- openTempFile "." "temp"
    let
        index = read itemId
        items = B.lines content
        newItems = delete (items !! index) items
    B.hPutStr tempHandle $ B.unlines newItems
    hClose tempHandle
    removeFile "todo.txt"
    renameFile tempName "todo.txt"

File: python/src/todo.py
#!/usr/bin/python

import os
import sys

def add(x):
    f = file("todo.txt", 'a')
    f.write(x + os.linesep)
    f.close()

def view():
    try:
        with open("todo.txt"): pass
    except IOError:
        print "todo.txt doesn't exist. Add something first?"
        return None
    f = file("todo.txt", 'r')
    i = 1
    while True:
        line = f.readline()
        if len(line) == 0:
            break
        print "%d - %s" % (i, line),
        i += 1
    f.close()

def rm(i):
    try:
        with open("todo.txt"): pass
    except IOError:
        print "todo.txt doesn't exist. Add something first?"
        return None
    items = []
    f = file("todo.txt", 'r')
    while True:
        line = f.readline()
        if len(line) == 0:
            break
        items.append(line)
    f.close()
    if (i > len(items)):
        print "Invalid index"
        return None
    del items[i - 1]
    f = file("todo.txt", 'w')
    for item in items:
        f.write(item)
    f.close()

print sys.version
argc = len(sys.argv)
if argc == 1:
    print "missing command"
    sys.exit(-1)
content = ""
if sys.argv[1] == "add":
    if argc == 2:
        print "missing argument"
        sys.exit(-2)
    for i in range(2, argc):
        content += (sys.argv[i] + " ")
    add(content)
elif sys.argv[1] == "view":
    view()
elif sys.argv[1] == "del":
    if argc == 2:
        print "missing argument"
        sys.exit(-2)
    index = int(sys.argv[2])
    rm(index)
else:
    print "unkown command"
    sys.exit(-3)
